[axonius][identity] Add Axonius Identity datastream

muskan-agarwal26 opened the pull request:
## Proposed commit message

The release includes identity data stream and associated dashboard.

Axonius fields are mapped to their corresponding ECS fields where possible.

Test samples were derived from live data samples, which were subsequently
sanitized.

## Checklist

- [x] I have reviewed [tips for building integrations](https://github.com/elastic/integrations/blob/main/docs/tips_for_building_integrations.md) and this pull request is aligned with them.
- [x] I have verified that all data streams collect metrics or logs.
- [x] I have added an entry to my package's `changelog.yml` file.
- [x] I have verified that Kibana version constraints are current according to [guidelines](https://github.com/elastic/elastic-package/blob/master/docs/howto/stack_version_support.md#when-to-update-the-condition).
- [x] I have verified that any added dashboard complies with Kibana's [Dashboard good practices](https://docs.elastic.dev/ux-guidelines/data-viz/dashboard-good-practices) 

## How to test this PR locally

To test the axonius package:

- Clone integrations repo.
- Install elastic package locally.
- Start elastic stack using elastic-package.
- Move to integrations/packages/axonius directory.
- Run the following command to run tests.

> elastic-package test
```
2025/12/18 17:48:46  INFO New version is available - v0.117.1. Download from: https://github.com/elastic/elastic-package/releases/tag/v0.117.1
Run asset tests for the package
2025/12/18 17:48:46  INFO License text found in "/root/GITHUB/integrations/LICENSE.txt" will be included in package
--- Test results for package: axonius - START ---
╭─────────┬─────────────┬───────────┬──────────────────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE │ DATA STREAM │ TEST TYPE │ TEST NAME                                                        │ RESULT │ TIME ELAPSED │
├─────────┼─────────────┼───────────┼──────────────────────────────────────────────────────────────────┼────────┼──────────────┤
│ axonius │             │ asset     │ dashboard axonius-60472232-ca7b-45e6-9fa6-72e6efc41a8e is loaded │ PASS   │      1.422µs │
│ axonius │ identity    │ asset     │ index_template logs-axonius.identity is loaded                   │ PASS   │        348ns │
│ axonius │ identity    │ asset     │ ingest_pipeline logs-axonius.identity-0.1.0 is loaded            │ PASS   │        233ns │
╰─────────┴─────────────┴───────────┴──────────────────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: axonius - END   ---
Done
Run pipeline tests for the package
--- Test results for package: axonius - START ---
╭─────────┬─────────────┬───────────┬─────────────────────────────────────────────────────┬────────┬──────────────╮
│ PACKAGE │ DATA STREAM │ TEST TYPE │ TEST NAME                                           │ RESULT │ TIME ELAPSED │
├─────────┼─────────────┼───────────┼─────────────────────────────────────────────────────┼────────┼──────────────┤
│ axonius │ identity    │ pipeline  │ (ingest pipeline warnings test-account.log)         │ PASS   │ 778.105629ms │
│ axonius │ identity    │ pipeline  │ (ingest pipeline warnings test-certificate.log)     │ PASS   │ 700.415371ms │
│ axonius │ identity    │ pipeline  │ (ingest pipeline warnings test-group.log)           │ PASS   │ 662.882359ms │
│ axonius │ identity    │ pipeline  │ (ingest pipeline warnings test-identity-common.log) │ PASS   │ 763.614554ms │
│ axonius │ identity    │ pipeline  │ (ingest pipeline warnings test-user.log)            │ PASS   │ 639.719729ms │
│ axonius │ identity    │ pipeline  │ test-account.log                                    │ PASS   │ 296.335698ms │
│ axonius │ identity    │ pipeline  │ test-certificate.log                                │ PASS   │ 309.327719ms │
│ axonius │ identity    │ pipeline  │ test-group.log                                      │ PASS   │ 308.939347ms │
│ axonius │ identity    │ pipeline  │ test-identity-common.log                            │ PASS   │ 291.192435ms │
│ axonius │ identity    │ pipeline  │ test-user.log                                       │ PASS   │ 316.869779ms │
╰─────────┴─────────────┴───────────┴─────────────────────────────────────────────────────┴────────┴──────────────╯
--- Test results for package: axonius - END   ---
Done
Run policy tests for the package
--- Test results for package: axonius - START ---
No test results
--- Test results for package: axonius - END   ---
Done
Run static tests for the package
--- Test results for package: axonius - START ---
╭─────────┬─────────────┬───────────┬──────────────────────────┬────────┬──────────────╮
│ PACKAGE │ DATA STREAM │ TEST TYPE │ TEST NAME                │ RESULT │ TIME ELAPSED │
├─────────┼─────────────┼───────────┼──────────────────────────┼────────┼──────────────┤
│ axonius │ identity    │ static    │ Verify sample_event.json │ PASS   │ 268.878426ms │
╰─────────┴─────────────┴───────────┴──────────────────────────┴────────┴──────────────╯
--- Test results for package: axonius - END   ---
Done
Run system tests for the package
2025/12/18 17:48:59  INFO Installing package...
2025/12/18 17:48:59  INFO License text found in "/root/GITHUB/integrations/LICENSE.txt" will be included in package
2025/12/18 17:49:01  INFO Running test for data_stream "identity" with configuration 'default'
2025/12/18 17:49:09  INFO Setting up independent Elastic Agent...
2025/12/18 17:49:20  INFO Setting up service...
2025/12/18 17:49:40  INFO Validating test case...
2025/12/18 17:49:42  INFO Tearing down service...
2025/12/18 17:49:43  INFO Write container logs to file: /root/GITHUB/integrations/build/container-logs/axonius-1766060383822768504.log
2025/12/18 17:49:47  INFO Tearing down agent...
2025/12/18 17:49:48  INFO Write container logs to file: /root/GITHUB/integrations/build/container-logs/elastic-agent-1766060388442966883.log
2025/12/18 17:50:15  INFO Uninstalling package...
--- Test results for package: axonius - START ---
╭─────────┬─────────────┬───────────┬───────────┬────────┬───────────────╮
│ PACKAGE │ DATA STREAM │ TEST TYPE │ TEST NAME │ RESULT │  TIME ELAPSED │
├─────────┼─────────────┼───────────┼───────────┼────────┼───────────────┤
│ axonius │ identity    │ system    │ default   │ PASS   │ 41.623316269s │
╰─────────┴─────────────┴───────────┴───────────┴────────┴───────────────╯
--- Test results for package: axonius - END   ---
Done
``` 

## Related issues

- Relates https://github.com/elastic/integrations/issues/15034
- Relates https://github.com/elastic/elasticsearch/pull/139512

## Screenshots
<img width="1919" height="913" alt="image (5)" src="https://github.com/user-attachments/assets/c254d523-7517-4137-a726-84e7c77f34fa" />
<img width="1919" height="916" alt="image (6)" src="https://github.com/user-attachments/assets/9abed0f7-58f5-4b4e-aca9-474d6de7b50f" />

muskan-crest committed:
Added identity datastream in Axonius

muskan-crest committed:
Update changelog entry

muskan-crest committed:
Added transform pipeline

muskan-crest committed:
Merge branch 'feature/axonius-0.1.0' of github.com:muskan-agarwal26/integrations into datastream-identity

efd6 commented on packages/axonius/data_stream/identity/agent/stream/cel.yml.hbs:
```diff
@@ -0,0 +1,140 @@
+                    :
+                      string(resp.Status) + " (" + string(resp.StatusCode) + ")"
+                  ),
+                  "asset_type": string(state.worklist.asset_type_list[0]),
```
Why does this need a `string` conversion?

muskan-agarwal26 replied:
No need, removing `string` conversion for this from everywhere in code.

efd6 commented on packages/axonius/_dev/deploy/docker/files/config.yml:
@muskan-agarwal26 Can you check this please. I don't think it's significant; it's true, but the pagination doesn't depend on it.

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,680 @@
+      target_field: axonius.identity.event.accurate_for_datetime
+      formats:
+        - EEE, dd MMM yyyy HH:mm:ss 'GMT'
+        - yyyy-MM-dd
+        - EEE,dd MMM yyyy HH:mm:ss 'GMT'
+      if: ctx.axonius?.identity?.event?.accurate_for_datetime != null && ctx.axonius.identity.event.accurate_for_datetime != ''
+      on_failure:
+        - append:
+            field: error.message
+            value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag {{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.on_failure_pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
```
Check that all `date` and `convert` processors will remove the field value on failure. This field is defined as a `date` so if it fails we have left an unmappable value in place.

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
Can this be more concisely/manageably expressed with some/most of the processors replaced with a `script`

efd6 commented on packages/axonius/changelog.yml:
```diff
@@ -1,4 +1,9 @@
+  changes:
+    - description: Add support for identity data stream.
+      type: enhancement
+      link: https://github.com/elastic/integrations/pull/16620
```
This should be in the 0.1.1 changes. This is not being merged into main in this PR.

efd6 replied:
Where did the changelog go?

muskan-agarwal26 replied:
there is no version change currently as you said in changelog hence kept it the same as earlier one's.

efd6 replied:
There is no version change, but there is a change description.

efd6 replied:
Please send another PR to add that since this has been merged now.

muskan-agarwal26 replied:
@efd6 , raised another PR: https://github.com/elastic/integrations/pull/18839

efd6 replied:
That change is not right; it should not *replace* the existing description, it should *add*:
```
    - description: Add support for identity data stream.
      type: enhancement
      link: https://github.com/elastic/integrations/pull/16620
```

efd6 commented on packages/axonius/manifest.yml:
```diff
@@ -1,10 +1,11 @@
 name: axonius
 title: Axonius
-version: 0.1.0
+version: 0.1.1
```
```suggestion
version: 0.1.0
```

efd6 commented on packages/axonius/kibana/dashboard/axonius-60472232-ca7b-45e6-9fa6-72e6efc41a8e.json:
```diff
@@ -0,0 +1,1343 @@
+    "id": "axonius-60472232-ca7b-45e6-9fa6-72e6efc41a8e",
+    "references": [
+        {
+            "id": "metrics-*",
```
logs?

efd6 commented on packages/axonius/kibana/dashboard/axonius-60472232-ca7b-45e6-9fa6-72e6efc41a8e.json:
```diff
@@ -0,0 +1,1343 @@
+            "type": "index-pattern"
+        },
+        {
+            "id": "metrics-*",
```
logs? Also, duplicate of the index-pattern at L1275.

efd6 commented on packages/axonius/kibana/dashboard/axonius-60472232-ca7b-45e6-9fa6-72e6efc41a8e.json:
```diff
@@ -0,0 +1,1343 @@
+            "name": "kibanaSavedObjectMeta.searchSourceJSON.filter[0].meta.index",
+            "type": "index-pattern"
+        },
+        {
+            "id": "logs-*",
+            "name": "kibanaSavedObjectMeta.searchSourceJSON.filter[1].meta.index",
+            "type": "index-pattern"
+        }
```
Duplicate of pattern at L1279.

efd6 commented on packages/axonius/elasticsearch/transform/latest_identity/fields/fields.yml:
```diff
@@ -0,0 +1,1392 @@
+              description: Remote identifier of the group.
+        - name: groups
+          type: keyword
+          description: List of group names this identity belongs to.
```
@muskan-agarwal26 Check if we need to use `subobjects: false` here. Do we have a system test that has both types of this field being populated?

muskan-agarwal26 replied:
This was a mistake, there should be only one definition as `group` type of `groups` key. Removing the duplicate one defined as `keyword`.

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,680 @@
+      tag: split_event_action
+      separator: \s+
+      ignore_missing: true
+      if: ctx.event?.action != ''
```
The processor also has an `ignore_missing: true`, so unless there is a present `null` there will not be a case that a `ctx.event?.action != ''` fires because of the LHS resolving to `null`.

muskan-crest committed:
Resolved comments provided by bot and @efd6.

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/default.yml:
```diff
@@ -0,0 +1,704 @@
+      name: '{{ IngestPipeline "pipeline-account" }}'
+      tag: pipeline-account
+      if: >-
+        ctx.axonius?.identity?.asset_type.contains('accounts')
```
This assessment is correct; there is no previous condition or action that guarantees that `asset_type` is not absent or null. An alternative fix is to use `ctx.axonius?.identity?.asset_type?.contains('accounts') == true`.

efd6 replied:
Leave all fixes for the author.

muskan-agarwal26 replied:
Keeping it as is.

efd6 replied:
@muskan-agarwal26 This should be fixed.

efd6 replied:
This has not been fixed.

efd6 commented on packages/axonius/elasticsearch/transform/latest_identity/fields/fields.yml:
```diff
@@ -0,0 +1,1413 @@
+              description: CIS benchmark version of the incompliant rule.
+            - name: rule_section
+              type: float
+              description: Section number of the incompliant CIS rule.
```
Agree.

efd6 commented on packages/axonius/elasticsearch/transform/latest_identity/fields/fields.yml:
```diff
@@ -0,0 +1,1413 @@
+              type: boolean
+              description: Indicates if cloud meeting recordings are auto-deleted.
+            - name: auto_delete_cmr_days
+              type: boolean
+              description: Indicates if auto-delete days for cloud recordings is configured.
```
Do we have a test that exercises this? Is it documented somewhere? I don't see it in any tests.

efd6 replied:
The comment was actually directed to @muskan-agarwal26.

muskan-agarwal26 replied:
yes, in live logs, we are getting value as `true` or `false`, hence it is kept boolean

efd6 replied:
@muskan-agarwal26 Please make it so that this is exercised in our pipeline tests.

efd6 replied:
This is not in tests yet.

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-group.yml:
```diff
@@ -0,0 +1,112 @@
+            - append:
+                field: error.message
+                value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag {{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.on_failure_pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
+  - foreach:
+      field: axonius.identity.user_count_link
+      tag: foreach_user_count_link_value
+      if: ctx.axonius?.identity?.user_count_link instanceof List
+      processor:
+        append:
+          field: related.user
+          tag: append_user_count_link_value_into_related_user
+          value: '{{{_ingest._value.value}}}'
+          allow_duplicates: false
```
This is not possible; conditionals in foreach processors are not usable. See [ref](https://github.com/elastic/elasticsearch/issues/60470).

muskan-agarwal26 replied:
any empty value will be dropped including the object in which it populates, so there is no need of if condition here

efd6 replied:
Resolve this.

efd6 commented on packages/axonius/kibana/dashboard/axonius-60472232-ca7b-45e6-9fa6-72e6efc41a8e.json:
```diff
@@ -0,0 +1,2343 @@
+            "name": "c6710940-9337-4981-9769-1fc79e4ece1a:009cf6e6-9a5a-4eb6-a90a-842d200df069",
+            "type": "index-pattern"
+        },
+        {
+            "id": "logs-*",
+            "name": "kibanaSavedObjectMeta.searchSourceJSON.filter[0].meta.index",
+            "type": "index-pattern"
+        },
+        {
+            "id": "logs-*",
+            "name": "kibanaSavedObjectMeta.searchSourceJSON.filter[1].meta.index",
+            "type": "index-pattern"
+        }
```
This duplication still exists.

muskan-agarwal26 replied:
I had reverted that but after exporting the dashboard again, the entry got added again
also, I had a look in other integration's dashboards, they have duplicate entries similarly

efd6 replied:
OK

efd6 reviewed (commented):
Please address https://github.com/elastic/integrations/pull/16620/changes#r3107948170 and the macroscopeapp concerns.

muskan-agarwal26 commented:
@efd6 I have addressed the comment:  https://github.com/elastic/integrations/pull/16620/changes#r3107948170 , used script processors to reduce the count of processors.

muskan-crest committed:
Resolved further comments of @efd6 and bot.

efd6 commented:
> I have addressed the comment: https://github.com/elastic/integrations/pull/16620/changes#r3107948170 , used script processors to reduce the count of processors.

I don't see any change in the size of that file.

muskan-agarwal26 commented:
> > I have addressed the comment: https://github.com/elastic/integrations/pull/16620/changes#r3107948170 , used script processors to reduce the count of processors.
> 
> I don't see any change in the size of that file.

It is actually reduced, it can be viewed in this commit: https://github.com/elastic/integrations/pull/16620/changes/8b75b008bd506c18330ed31744f20d932346c5f2

muskan-agarwal26 commented:
@efd6 , please have a look at my comment: https://github.com/elastic/integrations/pull/16620#issuecomment-4303052138

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -171,91 +171,36 @@ processors:
-                field: error.message
-                value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag {{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.on_failure_pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
+      source: |-
+        def fields = ['is_fabricated', 'is_retired', 'is_sensitive', 'is_spam_list', 'is_verified'];
```
Put this in a `params` array to avoid allocations.

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -171,91 +171,36 @@ processors:
+            def s = v.toString().toLowerCase();
+            if (s == 'true' || s == '1') {
+              item[f] = true;
+            } else if (s == 'false' || s == '0' || s == '') {
```
`if (s == '1' || s == 'true') {`

and

`} else if (s == '' || s == '0' || s == 'false') {`

(cheapest to most expensive)

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -416,75 +364,47 @@ processors:
+      description: Converts feature boolean and long fields, removing values that cannot be converted.
+      if: ctx.axonius?.identity?.feature != null
+      source: |-
+        def boolFields = ['cn_meeting', 'in_meeting', 'large_meeting', 'webinar', 'zoom_phone'];
```
Put this in a `params` array.

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -416,75 +364,47 @@ processors:
+          def s = v.toString().toLowerCase();
+          if (s == 'true' || s == '1') {
+            obj[f] = true;
+          } else if (s == 'false' || s == '0' || s == '') {
```
`if (s == '1' || s == 'true') {`

and

`} else if (s == '' || s == '0' || s == 'false') {`

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -503,634 +423,241 @@ processors:
+          'remote_control', 'remote_support', 'share_dual_camera',
+          'show_meeting_control_toolbar', 'virtual_background', 'waiting_room',
+          'workplace_by_facebook'
+        ];
```
`params` array.

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -503,634 +423,241 @@ processors:
+          'is_mfa_enrolled', 'is_non_editable', 'is_paid', 'is_permission_adapter',
+          'is_saas_user', 'is_user_active', 'is_user_deleted', 'is_user_external',
+          'is_user_inactive', 'is_user_suspended'
+        ];
```
`params` array.

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -503,634 +423,241 @@ processors:
+          def s = v.toString().toLowerCase();
+          if (s == 'true' || s == '1') {
+            obj[f] = true;
+          } else if (s == 'false' || s == '0' || s == '') {
```
order

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -503,634 +423,241 @@ processors:
+          def s = v.toString().toLowerCase();
+          if (s == 'true' || s == '1') {
+            obj[f] = true;
+          } else if (s == 'false' || s == '0' || s == '') {
```
order

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -1186,199 +713,89 @@ processors:
+          'host_pause_stop_recording', 'local_recording', 'record_audio_file',
+          'record_gallery_view', 'record_speaker_view', 'recording_audio_transcript',
+          'save_chat_text', 'show_timestamp'
+        ];
```
`params`

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -1186,199 +713,89 @@ processors:
+          def s = v.toString().toLowerCase();
+          if (s == 'true' || s == '1') {
+            obj[f] = true;
+          } else if (s == 'false' || s == '0' || s == '') {
```
order

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -1394,135 +811,38 @@ processors:
+          'require_password_for_pmi_meetings', 'require_password_for_scheduled_meetings',
+          'require_password_for_scheduling_new_meetings', 'use_pmi_for_instant_meetings',
+          'use_pmi_for_scheduled_meetings'
+        ];
```
`params`

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -1394,135 +811,38 @@ processors:
+          def s = v.toString().toLowerCase();
+          if (s == 'true' || s == '1') {
+            obj[f] = true;
+          } else if (s == 'false' || s == '0' || s == '') {
```
order

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -1615,162 +938,46 @@ processors:
+          'active_from_direct_adapter', 'is_from_direct_adapter', 'is_managed',
+          'is_saas_application', 'is_unmanaged_extension', 'is_user_deleted',
+          'is_user_external', 'is_user_paid', 'is_user_suspended'
+        ];
```
`params`

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -1615,162 +938,46 @@ processors:
+            def s = v.toString().toLowerCase();
+            if (s == 'true' || s == '1') {
+              item[f] = true;
+            } else if (s == 'false' || s == '0' || s == '') {
```
order

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -0,0 +1,1209 @@
+          field: related.user
+          tag: append_associated_groups_display_name_into_related_user
+          value: '{{{_ingest._value.display_name}}}'
+          allow_duplicates: false
```
This one as well.

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -0,0 +1,1209 @@
+          on_failure:
+            - remove:
+                field: _ingest._value.breach_date
+                ignore_missing: true
```
And this.

efd6 commented on packages/axonius/data_stream/identity/elasticsearch/ingest_pipeline/pipeline-user.yml:
```diff
@@ -0,0 +1,1209 @@
+                ignore_missing: true
+            - append:
+                field: error.message
+                value: 'Processor {{{_ingest.on_failure_processor_type}}} with tag {{{_ingest.on_failure_processor_tag}}} in pipeline {{{_ingest.on_failure_pipeline}}} failed with message: {{{_ingest.on_failure_message}}}'
```
… in conjunction with this which is also in `breaches_data` and so could be rolled in.

muskan-crest committed:
Resolved comments provided by @efd6 -
1. Adding params object inside script processors.
2. Adding more script to replace repeating foreach.
3. Added missing fields in tests.

efd6 reviewed (approved):
Thanks
